### PR TITLE
Clean up flake.nix

### PR DESCRIPTION
**Pull request by bloxx12**
This PR gets rid of the flake's dependency on `flake-utils`, which is an  abstraction library which is definitely not necessary here. This might be nice to reduce the flake's complexity and make it more accessible for non-`flake-utils` users. 

Furthermore, I've moved the `devShell` and `steel package` parts of the flake into seperate files, which has several benefits:
- Upstreaming steel into `Nixpkgs` becomes much easier
- Common in many projects which have a `flake.nix` 
- Reduces the flake's complexity by a bit

I still have to test this to make sure everything behaves the same way as before, but it should be more or less correct.

This is a draft, so I'm open to any changes ofc  :)

### CODE REVIEW

**Comment by mattwparas** [comment]
Thanks for this! Unfortunately I don't use nix - any recommendations on how to test this?

**Comment by bloxx12** [comment]
Hey, sorry for the late reply. I just tested everything and it should all just work. :)

**Comment by mattwparas** [comment]
> Hey, sorry for the late reply. I just tested everything and everything works properly :)

Great, no problem. What should I run to test this on my end? Also, is there a way I can test in the CI?

**Comment by bloxx12** [comment]
The original flake.nix exposes several things, namely:
1. A formatter
`      formatter = pkgs.alejandra;
`
2. The steel package
`      packages.steel = steel;
`
3. a default package (steel)
`      defaultPackage = packages.steel;
`
4. a devShell
`      devShell = with pkgs;
...
`
 5. a steel app
 `      apps.steel = {
        type = "app";
        program = "${steel}/bin/steel";
      };
`
 6. a default app (steel)
 `      apps.default = apps.steel;
`
The flake-utils library provides an abstraction over different architectures to make it a bit easier to provide all this for all architectures, although this can be just done like I've done here, using eachSystem:
alejandra is one of the standard nix formatters.
`    formatter = eachSystem (system: pkgsFor.${system}.alejandra);
`
which evaluates to:
```nix
formatter = {
	x86_64-linux = nixpkgs.legacyPackages.x86_64-linux.alejandra
	aarch64-linux = nixpkgs.legacyPackages.aarch64-linux.alejandra
	x86_64-darwin = nixpkgs.legacyPackages.x86_64-darwin.alejandra
	aarch64-darwin = nixpkgs.legacyPackages.aarch64-darwin.alejandra
```
	
formatter

I've done this for basically everything this flake exposes.
`legacyPackages = packages;` is basically redundant here but I've kept it to not change too much.

Now, to test these changes, there are several things which need to be tested:
(in the steel directory:)
- `nix build .` (builds the default package)
- `nix build .#steel` (builds the steel package)
- `nix run .` (builds and runs the default package)
- `nix run .#steel` (builds and runs the steel package)
- `nix develop .` (creates a devShell) 

I've sucessfully ran all of them and you should be able to too if you have nix installed.

As for the CI, I'm positive that's possible. I'll research a bit and comment later again. 
If you have any more questions I'd be glad to answer them :)

**Review by Dragyx — APPROVED** [review]
Seems to do exactly what it did before, just without flake-utils.
Separating the package and dev shell code into separate files is nice as well.

File: flake.nix [hunk 1]
```diff
@@ -1,87 +1,40 @@
...
   outputs = {
     self,
     nixpkgs,
-    flake-utils,
-  }:
-    flake-utils.lib.eachDefaultSystem (system: let
-      pkgs = nixpkgs.legacyPackages.${system};
-      fs = pkgs.lib.fileset;
+  }: let
+    systems = ["x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin"];
```
Comment (fufexan) [inline flake.nix:10]: Add `inputs.systems.url = "github:nix-systems/default";` and use `systems = import systems;` instead.

This way users can override systems to try and build on other platforms.

File: flake.nix [hunk 2]
```diff
@@ -1,87 +1,40 @@
...
+    packages = eachSystem (system: {
+      default = self.packages.${system}.steel;
+      steel = pkgsFor.${system}.callPackage ./nix/package.nix {
+        inherit (pkgsFor.${system}.darwin.apple_sdk.frameworks) Security;
+      };
+    });
+
+    formatter = eachSystem (system: pkgsFor.${system}.alejandra);
+    legacyPackages = self.packages;
+    defaultPackage = eachSystem (system: self.packages.${system}.default);
```
Comment (fufexan) [inline flake.nix:23]: Add a comment above these two that says `# DEPRECATED`. `legacyPackages` doesn't seem to need to be here, as the package set is flat, so it's already supported by `packages`.

`defaultPackage` is superseded by `packages.default`, which you already added.

File: flake.nix [hunk 3]
```diff
@@ -1,87 +1,40 @@
...
+        inherit (self.packages.${system}) steel;
+        inherit (pkgsFor.${system}.darwin.apple_sdk.frameworks) CoreServices SystemConfiguration;
+      };
+    });
+
+    apps = eachSystem (system: {
+      steel = {
         type = "app";
-        program = "${steel}/bin/steel";
+        program = "${self.packages.${system}.steel}/bin/steel";
```
Comment (fufexan) [inline flake.nix:35]:
```suggestion
        program = lib.getExe self.packages.${system}.steel;
```

File: nix/package.nix
Comment (fufexan) [inline nix/package.nix:1]: Newlines between each major binding in this file could go a long way in improving readability.
e.g. `pname` and `version` should be separated from `src` by a newline, same with `*inputs`.

File: nix/package.nix [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      substituteInPlace cogs/installer/download.scm --replace-warn "cargo-steel-lib" "$out/bin/cargo-steel-lib"
+      mkdir $out/lib
+      export STEEL_HOME="$out/lib"
+      pushd cogs
+      $out/bin/steel install.scm
+      popd
+      rm "$out/bin/cargo-steel-lib"
+    '';
+    meta = {
+      description = "An embedded scheme interpreter in Rust";
```
Comment (fufexan) [inline nix/package.nix:50]:
```suggestion
      description = "Embedded scheme interpreter in Rust";
```

File: nix/package.nix [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      homepage = "https://github.com/mattwparas/steel";
+      license = with lib.licenses; [asl20 mit];
+      mainProgram = "steel";
```
Comment (fufexan) [inline nix/package.nix:53]: Could also add `platforms = lib.platforms.unix;`

File: nix/shell.nix [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    [cargo rustc openssl libiconv]
+    ++ lib.optionals stdenv.isDarwin [
+      CoreServices
+      SystemConfiguration
+    ];
+  nativeBuildInputs = [
+    pkg-config
+    rust-analyzer
+    rustfmt
+  ];
```
Comment (fufexan) [inline nix/shell.nix:30]:
```suggestion
  packages = [
    cargo
    rustc
    rust-analyzer
    rustfmt
    libiconv
  ] ++ lib.optionals stdenv.isDarwin [
    CoreServices
    SystemConfiguration
  ];
  
  inputsFrom = steel;
```

`packages` should be used in shells, as the logical difference between `nativeBuildInputs` and `buildInputs` no longer applies.

`inputsFrom` takes pre-existing `nativeBuildInputs` and `buildInputs` from the given package and applies them to the shell.

**Review by fufexan — APPROVED** [review]

**Review by mattwparas — APPROVED** [review]

**Comment by mattwparas** [comment]
Thank you all for this; I don't daily drive nix so it's helpful to have others testing this regularly and improving. Perhaps at some point I'll start using nix 😄

**Pull request merged**